### ISSUE TITLE
use of closed network connection

### ISSUE THREAD

**Issue by Wollac** [go-zeromq/zmq4#39]
After disconnecting one of several SUBs, PUB starts returning the following error on Send:
```zmq4: error sending frame 1/1: write tcp 127.0.0.1:1234->127.0.0.1:58966: use of closed network connection```
Effectively, the server needs to be restarted after one client has been disconnected.

A minimal Go example for this would be the following, but it also happens with clients in other languages:
```
package main

import (
	"context"
	zmq "github.com/go-zeromq/zmq4"
	"log"
	"time"
)

// Send continuously
func send() {
	sock := zmq.NewPub(context.Background())
	defer sock.Close()

	if err := sock.Listen("tcp://*:1234"); err != nil {
		log.Fatal(err)
	}

	for {
		if err := sock.Send(zmq.NewMsgString("test")); err != nil {
			log.Println(err)
		}

		time.Sleep(time.Second)
	}
}

// Receive once
func receive() {
	sub := zmq.NewSub(context.Background())
	defer sub.Close()

	err := sub.Dial("tcp://localhost:1234")
	if err != nil {
		log.Fatal(err)
	}

	err = sub.SetOption(zmq.OptionSubscribe, "test")
	if err != nil {
		log.Fatal(err)
	}

	msg, err := sub.Recv()
	if err != nil {
		log.Fatal(err)
	}
	log.Println(msg)
}

func main() {
	go send()
	receive() // receive one and close
	time.Sleep(5 * time.Second)
}
```

**Comment by sbinet** [comment]
apologies for the belated answer.
I am currently a bit stretched thin across various projects of mine.

I am afraid I won't be able to dedicate much time until end of July/beg of August.
sorry.

PRs welcome, though.

**Comment by bjasionowski** [comment]
I encountered the same issue w/ the request reply messaging paradigm. It appears when a client connects to the server, it creates a connection/message writer but it never closes it when the client disconnects. As such subsequent connections encounter this error. I instrumented msgio.go and observe the following:
``
[server] 2019/09/24 11:39:32 Setup the logger
[server] 2019/09/24 11:39:32 Serving on port 1234
Adding writer for connection
[server] 2019/09/24 11:39:34 Received EchoRequest
[server] 2019/09/24 11:39:34 Echo message: 'I believe I can fly'
Sending 1 of 1, with message: [[] [1010 1101 1000 11010110 11011100 11010101 10010001 11111000 11111111 11111111 11111111 11111111 1 10000 1 10010 10101 11010 10011 1001001 100000 1100010 1100101 1101100 1101001 1100101 1110110 1100101 100000 1001001 100000 1100011 1100001 1101110
100000 1100110 1101100 1111001]]
[server] 2019/09/24 11:39:34 Sent 1 frame/s
Adding writer for connection
[server] 2019/09/24 11:39:40 Received EchoRequest
[server] 2019/09/24 11:39:40 Echo message: 'I believe I can fly'
Sending 1 of 2, with message: [[] [1010 1101 1000 11010110 11010110 10010100 10101110 11111000 11111111 11111111 11111111 11111111 1 10000 1 10010 10101 11010 10011 1001001 100000 1100010 1100101 1101100 1101001 1100101 1110110 1100101 100000 1001001 100000 1100011 1100001 1101110
100000 1100110 1101100 1111001]]
Sending 2 of 2, with message: [[] [1010 1101 1000 11010110 11010110 10010100 10101110 11111000 11111111 11111111 11111111 11111111 1 10000 1 10010 10101 11010 10011 1001001 100000 1100010 1100101 1101100 1101001 1100101 1110110 1100101 100000 1001001 100000 1100011 1100001 1101110
100000 1100110 1101100 1111001]]
[server] 2019/09/24 11:39:40 Unable to transmit reply over 1 frames: zmq4: error sending frame 1/2: write tcp 127.0.0.1:1234->127.0.0.1:41070: use of closed network connection
``
My client receives the message on the second trial, but the server encounters an error and crashes due to unclosed connection.
